_posts/2018-05-02-load-balanced-smart-proxies-with-haproxy.md

```diff
@@ -0,0 +1,151 @@
+
+## Requirements
+* an existing working Foreman/Katello installation (we don't need to do any changes here, setup is the normal installation)
+* two or more machines for the Smart Proxy cluster
+* a load balancer - you can use a machine with HAProxy if you don't already have a load balancer
+* a common name in the DNS that points to the load balancer (we'll be using `smartproxy.example.net`)
+
+## Installing the first Smart Proxy
+### Generating Katello certificates
+When calling `foreman-proxy-certs-generate`, we have to pass `--foreman-proxy-cname smartproxy.example.net` to add the load-balanced name as an accepted `subjectAltName` to the list of names in the generated certificates. This allows clients to connect to each member of the smartproxy cluster as `smartproxy.example.net`
+
+### Preparing `custom-hiera.yaml`
+Pulp uses the hostname of the smart proxy to generate redirects when serving [lazy (on demand) synced content](https://docs.pulpproject.org/dev-guide/design/deferred-download.html). These redirects need to happen to the name of the load-balancer, but this option is not exposed in the installer. Thus edit `/etc/foreman-installer/custom-hiera.yaml` and add the following line at the end:
+
+```yaml
+pulp::lazy_redirect_host: smartproxy.example.net
+```
+### Running the installer
+As only the first Smart Proxy will host a PuppetCA, we configure that explicitly during the installation and also configure the Puppet certificate to include the `smartproxy.example.net` name:
+```
+[root@smartproxy01] # foreman-installer \
+…
+--puppet-dns-alt-names smartproxy.example.net \
+--puppet-ca-server proxy01.example.net \
+--foreman-proxy-puppetca true \
+--puppet-server-ca true
+```
+
+## Installing the other Smart Proxies
+The installation of all other Smart Proxies is almost identical to the first one. We just have to prepare the Puppet certs on the first one (as it's the only one having a PuppetCA) and disable the PuppetCA in the installer.
+
+### Generating Katello certificates
+This step is identical to the steps used to generate the certificates for the first smart proxy. Simply use `foreman-proxy-certs-generate` as described above.
+
+### Generating Puppet certificates
+On the first smart proxy, generate the Puppet certificates by running
+
+```shell
+puppet cert generate proxy02.example.net --dns_alt_names=smartproxy.example.net
+```
+
+Install the `puppetserver` package on the second smart proxy, to create the `puppet` user and the directory structure in `/etc/puppetlabs`.
+
+Now copy `/etc/puppetlabs/puppet/ssl/*/proxy02.example.net.pem` and `/etc/puppetlabs/puppet/ssl/certs/ca.pem` to the second smart proxy and ensure ownership and permissions are identical to those the first host. Also don't forget to check the SELinux contexts of the files you copied over (hint: `ls -lZ`) :)
```

Comment:
I realize now you're already passing in the right vars in the installer. Not sure if it works, but I think you can add `proxy02.example.net` to `/etc/puppetlabs/puppet/autosign.conf` on `proxy01.example.net` and I think the installer should get the cert from its CA. Not sure though.

Reply:
I tried that while working on the forklift changes, and it somehow did not work. No idea why, but proxy02 never did a CA request to proxy01. Might re-test it at some point, but the above works.

Reply:
I think it does make sense, but we could have a look at supporting something like it. For now this works indeed :)